### PR TITLE
fix: Preserve line numbers when using she-bangs

### PR DESCRIPTION
**Pull request by samuelorji**
Hopefully fixes https://github.com/epage/cargo-script-mvs/issues/59

 
Screenshot 2022-08-06 at 22 40 18

Using the example in the issue, it now shows error on line 8.

### CODE REVIEW

**Comment by epage** [comment]
Thanks for looking into this!

Could you look into us leaving the she-bang in instead of removing it?  According to the reference, [she bangs are legal](https://doc.rust-lang.org/reference/crates-and-source-files.html)

File: src/manifest.rs [hunk 1]
```diff
@@ -26,12 +26,9 @@ static RE_SHEBANG: once_cell::sync::Lazy<Regex> =
...
 static RE_CRATE_COMMENT: once_cell::sync::Lazy<Regex> = once_cell::sync::Lazy::new(|| {
     Regex::new(
-            r"(?x)
-                # We need to find the first `/*!` or `//!` that *isn't* preceded by something that would make it apply to anything other than the crate itself.  Because we can't do this accurately, we'll just require that the doc comment is the *first* thing in the file (after the optional shebang, which should already have been stripped).
-                ^\s*
-                (/\*!|//(!|/))
-            "
-        ).unwrap()
+        // We need to find the first `/*!` or `//!` that *isn't* preceded by something that would make it apply to anything other than the crate itself.  Because we can't do this accurately, we'll just require that the doc comment is the *first* thing in the file (after the optional shebang).
+        r"(?x)(^\s*|^\#![^\[].*?(\r\n|\n))(/\*!|//(!|/))"
```
Comment (samuelorji) [inline src/manifest.rs:30]: add regex for optional shebang since we're not stripping it off anymore

File: src/manifest.rs [hunk 2]
```diff
@@ -51,14 +48,16 @@ pub fn split_input(input: &Input, input_id: &OsString) -> MainResult<(String, St
...
             let (manifest, source) =
                 find_embedded_manifest(content).unwrap_or((Manifest::Toml(""), content));
 
             let source = if source.lines().any(contains_main_method) {
                 source.to_string()
             } else {
-                format!("fn main() -> Result<(), Box<dyn std::error::Error+Sync+Send>> {{\n    {{\n    {}    }}\n    Ok(())\n}}", source)
+                let (content, found_shebang) = strip_shebang(content.as_str());
+                // use a newline separator when a shebang is found, and a tab when no shebang is found to preserve original line numbering
+                let separator = if found_shebang { "\n" } else { "\t" };
```
Comment (samuelorji) [inline src/manifest.rs:59]: We have to strip out shebang if there is no main method.
To preserve original line numbering, use a little hack where if the shebang is stripped, we replace with a new line or a tab if no shebang is found

File: src/manifest.rs [hunk 3]
```diff
@@ -467,7 +466,7 @@ fn find_code_block_manifest(s: &str) -> Option<(Manifest, &str)> {
     Then, we need to take the contents of this doc comment and feed it to a Markdown parser.  We are looking for *the first* fenced code block with a language token of `cargo`.  This is extracted and pasted back together into the manifest.
     */
     let start = match RE_CRATE_COMMENT.captures(s) {
-        Some(cap) => match cap.get(1) {
+        Some(cap) => match cap.get(3) {
```
Comment (samuelorji) [inline src/manifest.rs:469]: Due to regex change, we need to capture the 3rd group

File: tests/integration/script.rs [hunk 4]
```diff
@@ -1,4 +1,66 @@
...
+#[test]
+fn test_script_line_numbering_preserved_no_main() {
+    let fixture = crate::util::Fixture::new();
+    fixture
+        .cmd()
+        .arg("tests/data/script_line_numbering_preserved_no_main.rs")
+        .assert()
+        .success()
+        .stdout_eq(
+            "line: 3
```
Comment (samuelorji) [inline tests/integration/script.rs:26]: as seen [here](https://github.com/epage/cargo-script-mvs/blob/0f5185ad7edfbb8d36b239e01ef80452543f943e/tests/data/script_line_numbering_preserved_no_main.rs#L3)

File: tests/integration/script.rs [hunk 5]
```diff
@@ -1,4 +1,66 @@
...
+#[test]
+fn test_script_line_numbering_preserved_no_main_no_shebang() {
+    let fixture = crate::util::Fixture::new();
+    fixture
+        .cmd()
+        .arg("tests/data/script_line_numbering_preserved_no_main_no_shebang.rs")
+        .assert()
+        .success()
+        .stdout_eq(
+            "line: 2
```
Comment (samuelorji) [inline tests/integration/script.rs:57]: as seen [here](https://github.com/epage/cargo-script-mvs/blob/0f5185ad7edfbb8d36b239e01ef80452543f943e/tests/data/script_line_numbering_preserved_no_main_no_shebang.rs#L2)

File: tests/integration/script.rs [hunk 6]
```diff
@@ -1,4 +1,66 @@
...
+#[test]
+fn test_script_line_numbering_preserved_no_shebang() {
+    let fixture = crate::util::Fixture::new();
+    fixture
+        .cmd()
+        .arg("tests/data/script_line_numbering_preserved_no_shebang.rs")
+        .assert()
+        .success()
+        .stdout_eq(
+            "line: 11
```
Comment (samuelorji) [inline tests/integration/script.rs:41]: as seen [here](https://github.com/epage/cargo-script-mvs/blob/0f5185ad7edfbb8d36b239e01ef80452543f943e/tests/data/script_line_numbering_preserved_no_shebang.rs#L11)

File: tests/integration/script.rs [hunk 7]
```diff
@@ -1,4 +1,66 @@
 #[test]
+fn test_script_line_numbering_preserved() {
+    let fixture = crate::util::Fixture::new();
+    fixture
+        .cmd()
+        .arg("tests/data/script_line_numbering_preserved.rs")
+        .assert()
+        .success()
+        .stdout_eq(
+            "line: 12
```
Comment (samuelorji) [inline tests/integration/script.rs:10]: as seen [here](https://github.com/epage/cargo-script-mvs/blob/0f5185ad7edfbb8d36b239e01ef80452543f943e/tests/data/script_line_numbering_preserved.rs#L12)

**Review by samuelorji — COMMENTED** [review]
@epage is this an improvement?

**Review by epage — APPROVED** [review]
Thanks for providing comments along the way to help with the review!

**Comment by epage** [comment]
The windows tests are failing in master, so don't worry about it

**Comment by samuelorji** [comment]
> Details



> The windows tests are failing in master, so don't worry about it

Does that have to do with my change or it's a regular occurrence ?

**Comment by samuelorji** [comment]
Guess I didn't use the right commit structure,  guess it has to be `type: description`

**Comment by epage** [comment]
> > Details
> 
> > The windows tests are failing in master, so don't worry about it
> 
> Does that have to do with my change or it's a regular occurrence ?

Its failing in master which means its from before your changes.  I think its related to the last rustup release but I haven't had time to diagnose it

**Comment by epage** [comment]
@samuelorji whats the status of fixing the rustfmt issue?

**Comment by samuelorji** [comment]
@epage lemme look into it now

**Comment by samuelorji** [comment]
@epage I've run `committed` and `cargo fmt`, and all seems fine from my end, lemme see if this helps

**Pull request merged**